Fix upgrade tests with and without snapshot

npepinpe opened the pull request:
## Description

This PR introduces major changes to the upgrade test suite. The first is a rewrite to junit5, which allows us to split the main suite into different concerns, such that we now have the `UpgradeTest`, which has the individual tests (withSnapshot, withoutSnapshot, withOldGateway, etc.); `UpgradeTestCase`, which describes a single test case used in the parameterized tests of `UpgradeTest`; `UpgradeTestCaseProvider`, which generates all the test cases; `ContainerState`, which is a simple POJO representing the state of the containers used by the `UpgradeTest`; and `ContainerStateExtension`, which wraps the `ContainerState` and handles its lifecycle based on the junit5 test lifecycle hooks.

The second one is introducing a `PartitionsActuatorClient`, which allows us to leverage the broker admin endpoint to force taking snapshot, and poll for new snapshots, removing the need to check the filesystem for this (or wait some time). Unfortunately we still need to stop/start the container in order to disable the debug exporter so we can take a snapshot.

Finally, the PR removes the shared data folder in favour of reusable Docker volumes, which are properly managed by Testcontainers by adding the right labels, such that even if the JVM crashes, the Ryuk container will reap the volumes.

This PR should be backported once approved.

## Related issues

closes #5385 
closes #5804

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

npepinpe commented:
Note that this adds additional time to our build (~5 minutes, sometimes more), so I'd like to do what @Zelldon proposed and reuse @korthout's work on https://github.com/zeebe-io/zeebe/pull/5735 and split the Upgrade Tests from the integration tests and run them in a separate agent, which should help not only speed things up but make them more stable.

MiguelPires commented on test-util/src/main/java/io/zeebe/test/util/testcontainers/ManagedVolume.java:
```diff
@@ -0,0 +1,93 @@
+    try (final CreateVolumeCmd command = client.createVolumeCmd()) {
+      final CreateVolumeResponse response =
+          command.withLabels(DockerClientFactory.DEFAULT_LABELS).exec();
+      return new ManagedVolume(response.getName());
```
You said that volumes are "properly managed by Testcontainers by adding the right labels". Is this what you meant? I tried googling a bit on how it works but didn't find much

npepinpe replied:
Yes, this is undocumented unfortunately, but Sergei is the one who told me to use this. I imagine one downside is it's a little brittle if they ever change how Ryuk works, but I think that's a small thing. Of course it would be nice if Testcontainers already provided the feature, and maybe that's something I'll contribute at some point so we have an official, maintained API on how to do this.

MiguelPires commented on upgrade-tests/src/test/java/io/zeebe/test/UpgradeTest.java:
```diff
@@ -7,453 +7,134 @@
+      // it's necessary to restart without the debug exporter to allow snapshotting
       state.close();
+      state.broker(LAST_VERSION).start(false);
+      EitherAssert.assertThat(state.getPartitionsActuatorClient().takeSnapshot()).isRight();
```
I'm not adding a change request because it's just an opinion but I don't think this is very readable. The assert `isRight` doesn't look good to me but I don't have a good alternative so we can leave it. Maybe `hasRight` would also be misleading because it sounds like it can have both right and left. I think extracting the value would make it more obvious what the assert is actually checking though:
```suggestion
      Either<Throwable, Map<String, PartitionStatus>> response = state.getPartitionsActuatorClient().takeSnapshot();
      EitherAssert.assertThat(response).isRight();
```

npepinpe replied:
Hm, I think that's just a fundamental problem with a generic structure like `Either` =/ I admit in this case I'd rather have something like Rust's [Result](https://doc.rust-lang.org/std/result/) type, where you would then write something like `isOk()` (since it's either OK or failed, which is really what I use it as for the client).

npepinpe replied:
I added a description to the assert via `as()` for better error message :+1:

MiguelPires reviewed (approved):
Looks good. I'm approving but not merging since you might want to squash

npepinpe commented:
One change I added which seems big but isn't really - I renamed upgrade to update, as we discussed in the post mortem. But it's just a plain renaming, nothing special (i.e. just done via the IDE).

npepinpe commented:
bors r+
